src/aotcompile.cpp

```diff
@@ -0,0 +1,909 @@
+    jl_native_code_desc_t *data = (jl_native_code_desc_t*)native_code;
+    if (data) {
+        auto it = data->jl_value_to_llvm.find(p);
+        if (it != data->jl_value_to_llvm.end()) {
+            return it->second;
+        }
+    }
+    return 0;
+}
+
+extern "C" JL_DLLEXPORT
+Module* jl_get_llvm_module(void *native_code)
+{
+    jl_native_code_desc_t *data = (jl_native_code_desc_t*)native_code;
+    if (data)
+        return data->M.get();
+    else
+        return NULL;
+}
+
+extern "C" JL_DLLEXPORT
+GlobalValue* jl_get_llvm_function(void *native_code, uint32_t idx)
+{
+    jl_native_code_desc_t *data = (jl_native_code_desc_t*)native_code;
+    if (data)
+        return data->jl_sysimg_fvars[idx];
+    else
+        return NULL;
+}
+
+extern "C" JL_DLLEXPORT
+LLVMContext* jl_get_llvm_context(void *native_code)
+{
+    jl_native_code_desc_t *data = (jl_native_code_desc_t*)native_code;
+    if (data)
+        return &data->M->getContext();
+    else
+        return NULL;
+}
+
+
+static void emit_offset_table(Module &mod, const std::vector<GlobalValue*> &vars, StringRef name, Type *T_psize)
+{
+    // Emit a global variable with all the variable addresses.
+    // The cloning pass will convert them into offsets.
+    assert(!vars.empty());
+    size_t nvars = vars.size();
+    std::vector<Constant*> addrs(nvars);
+    for (size_t i = 0; i < nvars; i++) {
+        Constant *var = vars[i];
+        addrs[i] = ConstantExpr::getBitCast(var, T_psize);
+    }
+    ArrayType *vars_type = ArrayType::get(T_psize, nvars);
+    new GlobalVariable(mod, vars_type, true,
+                       GlobalVariable::ExternalLinkage,
+                       ConstantArray::get(vars_type, addrs),
+                       name);
+}
+
+static bool is_safe_char(unsigned char c)
+{
+    return ('0' <= c && c <= '9') ||
+           ('A' <= c && c <= 'Z') ||
+           ('a' <= c && c <= 'z') ||
+           (c == '_' || c == '.') ||
+           (c >= 128 && c < 255);
+}
+
+static const char hexchars[16] = {
+    '0', '1', '2', '3', '4', '5', '6', '7',
+    '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };
+
+static const char *const common_names[256] = {
+//  0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, e, f
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 0x00
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 0x10
+    0, "NOT", 0, "YY", "XOR", 0, "AND", 0, 0, 0, "MUL", "SUM", 0, "SUB", 0, "DIV", // 0x20
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, "COL", 0, "LT", "EQ", "GT", 0, // 0x30
+    "AT", 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 0x40
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, "RDV", 0, "POW", 0, // 0x50
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 0x60
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, "OR", 0, "TLD", 0, // 0x70
+    0 };
+
+// removes special characters from the name of GlobalObjects,
+// which might cause them to be treated special by LLVM
+// or the system linker
+// the only non-identifier characters we keep are '%' and '.',
```

Comment:
Could you use an other character? `%` isn't actually safe, and doesn't get past the NVPTX assembler: https://github.com/llvm/llvm-project/blob/aa324c5441f229bd1ff07407affeab95740328fe/llvm/lib/MC/MCAsmInfo.cpp#L103-L106

Reply:
OK. Alternates might be `$` or use a digram like `..` or be non-reversible and use no signaling character or use a pattern like `_AAB_`. Thoughts/preference?

Reply:
`$` seems like a good alternative, but I'll have to verify not only LLVM's NVPTX but also `ptxax` is happy with that.

Reply:
New version is nvptx/ptxas-compatible.